src/MongoDB.Driver/Encryption/LibMongoCryptController.cs

```diff
@@ -0,0 +1,681 @@
+            var results = cursor.ToList(cancellationToken);
+            FeedResults(context, results);
+        }
+
+        private async Task ProcessNeedCollectionInfoStateAsync(CryptContext context, string databaseName, CancellationToken cancellationToken)
+        {
+            var database = _client.GetDatabase(databaseName);
+            var filterBytes = context.GetOperation().ToArray();
+            var filterDocument = new RawBsonDocument(filterBytes);
+            var filter = new BsonDocumentFilterDefinition<BsonDocument>(filterDocument);
+            var options = new ListCollectionsOptions { Filter = filter };
+            var cursor = await database.ListCollectionsAsync(options, cancellationToken).ConfigureAwait(false);
+            var results = await cursor.ToListAsync(cancellationToken).ConfigureAwait(false);
+            FeedResults(context, results);
+        }
+
+        private void ProcessNeedKmsState(CryptContext context, CancellationToken cancellationToken)
+        {
+            var requests = context.GetKmsMessageRequests();
+            foreach (var request in requests)
+            {
+                SendKmsRequest(request, cancellationToken);
+            }
+            requests.MarkDone();
+        }
+
+        private async Task ProcessNeedKmsStateAsync(CryptContext context, CancellationToken cancellationToken)
+        {
+            var requests = context.GetKmsMessageRequests();
+            foreach (var request in requests)
+            {
+                await SendKmsRequestAsync(request, cancellationToken).ConfigureAwait(false);
+            }
+            requests.MarkDone();
+        }
+
+        private void ProcessNeedMongoKeysState(CryptContext context, CancellationToken cancellationToken)
+        {
+            var filterBytes = context.GetOperation().ToArray();
+            var filterDocument = new RawBsonDocument(filterBytes);
+            var filter = new BsonDocumentFilterDefinition<BsonDocument>(filterDocument);
+            var cursor = _keyVaultCollection.FindSync(filter, cancellationToken: cancellationToken);
+            var results = cursor.ToList(cancellationToken);
+            FeedResults(context, results);
+        }
+
+        private async Task ProcessNeedMongoKeysStateAsync(CryptContext context, CancellationToken cancellationToken)
+        {
+            var filterBytes = context.GetOperation().ToArray();
+            var filterDocument = new RawBsonDocument(filterBytes);
+            var filter = new BsonDocumentFilterDefinition<BsonDocument>(filterDocument);
+            var cursor = await _keyVaultCollection.FindAsync(filter, cancellationToken: cancellationToken).ConfigureAwait(false);
+            var results = await cursor.ToListAsync(cancellationToken).ConfigureAwait(false);
+            FeedResults(context, results);
+        }
+
+        private void ProcessNeedMongoMarkingsState(CryptContext context, string databaseName, CancellationToken cancellationToken)
```

Comment:
Do we need a try/catch here that throws a new exception type to satisfy this part of the spec:
[All errors from the MongoClient to mongocryptd MUST be distinguished in some way (e.g. exception type) to make it easier for users to distinguish when a command fails due to auto encryption limitations.](https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#implementation)

Reply:
I've done it via adding `try/catch`s for top-level public methods in this class. Do you think we need to add the same logic for each/some particular stages too?

Reply:
Hm... I think the spec specifies that only errors from `MongoClient` to `mongocryptd` need to be distinguished in some way: wouldn't this mean we would only need a try/catch for `ProcessNeedMongoMarkingsState`  Your approach of wrapping the original exception in a MongoClientException seems reasonable to me, although it might be a bit more user friendly to prepend the exception's message with `mongocryptd exception`.  Alternatively, we could create a new exception type (e.g. `MongoCryptDException`) while still wrapping the original exception.

Reply:
Discussed offline. I've added a new exception:
```
           catch (Exception ex)
            {
                throw new MongoEncryptionException($"Exception in encryption library: {ex.Message}.", ex);
            }
```